### PR TITLE
Add always seekable body plugin

### LINKED ISSUE #98: FilteredStream: unable to read content of stream multiple times
| Q            | A
| ------------ | ---
| Bug?         | yes
| New Feature? | no
| Version      | 1.7.0

I am working on a 3rd party integration for my library that uses HTTplug and this library. When I integrated my library to Drupal 8 that has its own Guzzle client wrapper I faced with an interesting problem. By using the `\Http\Client\Common\Plugin\DecoderPlugin` sometimes I got back an empty string instead of the JSON that API has actually returned.

I think this is the combination of two possible issues:

1.  If something reads the content of the full response body before the `\Http\Client\Common\Plugin\DecoderPlugin` plugin can do that then when it starts processing the response (ex.: with the `\Http\Message\Encoding\DechunkStream`) it does not rewinds the stream, therefore when the `\Http\Message\Encoding\FilteredStream::getContents()` is being called the while loop does not do anything because the wrapped (underlying) stream has already in the end.
Maybe this can be solved by adding this condition to the constructor of the FilteredStream class.

```php
if ($this->stream->isSeekable()) {
  $this->stream->rewind();
}
```

2. If the wrapped stream was being closed then any of the `\Http\Message\Encoding\FilteredStream` implementations could not read the content of the stream again. It means that if I use the `\Http\Client\Common\Plugin\DecoderPlugin` on my client an try to get the body of the response multiple times like this: `(string) $response->getBody()` then after the first call I always get back an empty string. (Or if we add the suggested condition to the FilteredStream class' constructor an exception, because the read() method tries to read on a closed stream after the first `(string) $response->getBody()`). If I remove the plugin then the originally wrapped `\GuzzleHttp\Psr7\Stream`  object inside an instance of `\GuzzleHttp\Psr7\Response` can always return the same response body (string) multiple times.

### PR DESCRIPTION
**Pull request by joelwurtz**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #98, replace #99, partially #97  
| Documentation   | https://github.com/php-http/documentation/pull/251
| License         | MIT

#### What's in this PR?

Add a plugin that ensure request body or response body is always seekable

#### Why?

This may be needed when multiple librairies need to acces the content but are not aware of each other, so this allow them to read the body once and then rewind it.

### CODE REVIEW

**Comment by joelwurtz** [comment]
Tests are green, don't know why travis ci report this as failed ^^

File: spec/Plugin/AlwaysSeekableBodyPluginSpec.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+    {
+        $this->shouldHaveType(AlwaysSeekableBodyPlugin::class);
+    }
+
+    public function it_is_a_plugin()
+    {
+        $this->shouldImplement(Plugin::class);
+    }
+
+    public function it_decorate_response_body_if_not_seekable(RequestInterface $request, ResponseInterface $response, StreamInterface $responseStream, StreamInterface $requestStream)
```
Comment (dbu) [inline spec/Plugin/AlwaysSeekableBodyPluginSpec.php]: we have issue #45 saying we should stop mocking request and response objects. i suggest to instantiate actual guzzlehttp/psr7 objects instead.

File: src/Plugin/AlwaysSeekableBodyPlugin.php [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+use Http\Client\Common\Plugin;
+use Http\Message\Stream\BufferedStream;
+use Http\Promise\Promise;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+
+/**
+ * Decorate the body of the request and the response if it's not seekable by using Http\Message\Stream\BufferedStream.
```
Comment (dbu) [inline src/Plugin/AlwaysSeekableBodyPlugin.php]: s/if it's not seekable/if they are not seekable/

**Comment by dbu** [comment]
hm, good question. it could make sense to distinguish that, because i assume adding the buffer is a significant cost that should be avoided when its unnecessary. rather than having flags, i would prefer to split the plugins into SeekableRequestBodyPlugin and SeekableResponseBodyPlugin 

(either way i think we can drop the "Always" from the name, it is more verbose without adding much more clarity imho)

File: spec/Plugin/AlwaysSeekableBodyPluginSpec.php (hunk 1 above)
Comment (joelwurtz) [reply spec/Plugin/AlwaysSeekableBodyPluginSpec.php]: i'm fine with #45 but here i need to test that some method on request / response are being (not) called, so i think it's better to keep the mocking here

Comment (dbu) [reply spec/Plugin/AlwaysSeekableBodyPluginSpec.php]: fair enough, then lets keep mocking.

**Comment by joelwurtz** [comment]
As you wish, we can also spli this into 2 plugins : RésponseBodySeekable and RequestBodySeekable, which may be better since you will not put this 2 plugins at the same place : 

 * RequestBodySeekable should be in first so following plugins can seek its body
 * ResponseBodySeekable should be in last so previous plugins will correct have a seekable body on the response

**Comment by dbu** [comment]
didn't even think of that, but yes, lets split the plugin then.

File: src/Plugin/AlwaysSeekableBodyPlugin.php (hunk 2 above)
Comment (joelwurtz) [reply src/Plugin/AlwaysSeekableBodyPlugin.php]: are you sure about that (it = the body) ? However does not matter if splitted :p

Comment (dbu) [reply src/Plugin/AlwaysSeekableBodyPlugin.php]: haha, you were right, i got confused. but yes, when we split its going to be singular on both :-)

**Comment by joelwurtz** [comment]
Should be good, will update the documentation accordingly if this is good to merge

**Pull request merged**